pkg/server/v2/server.go

```diff
@@ -136,381 +118,20 @@ func (c CallbackFuncs) OnFetchResponse(req *discovery.DiscoveryRequest, resp *di
-	return &server{cache: config, callbacks: callbacks, ctx: ctx}
+	return NewServerAdvanced(rest.NewServer(config, callbacks), sotw.NewServer(ctx, config, callbacks))
 }
 
-type server struct {
-	cache     cache.Cache
-	callbacks Callbacks
-
-	// streamCount for counting bi-di streams
-	streamCount int64
-	ctx         context.Context
-}
-
-// Generic RPC stream.
-type Stream interface {
-	grpc.ServerStream
-
-	Send(*discovery.DiscoveryResponse) error
-	Recv() (*discovery.DiscoveryRequest, error)
-}
-
-// watches for all xDS resource types
-type watches struct {
-	endpoints chan cache.Response
-	clusters  chan cache.Response
-	routes    chan cache.Response
-	listeners chan cache.Response
-	secrets   chan cache.Response
-	runtimes  chan cache.Response
-
-	endpointCancel func()
-	clusterCancel  func()
-	routeCancel    func()
-	listenerCancel func()
-	secretCancel   func()
-	runtimeCancel  func()
-
-	endpointNonce string
-	clusterNonce  string
-	routeNonce    string
-	listenerNonce string
-	secretNonce   string
-	runtimeNonce  string
-
-	// Opaque resources share a muxed channel. Nonces and watch cancellations are indexed by type URL.
-	responses     chan cache.Response
-	cancellations map[string]func()
-	nonces        map[string]string
-	terminations  map[string]chan struct{}
-}
-
-// Initialize all watches
-func (values *watches) Init() {
-	// muxed channel needs a buffer to release go-routines populating it
-	values.responses = make(chan cache.Response, 5)
-	values.cancellations = make(map[string]func())
-	values.nonces = make(map[string]string)
-	values.terminations = make(map[string]chan struct{})
-}
-
-// Token response value used to signal a watch failure in muxed watches.
-var errorResponse = &cache.RawResponse{}
-
-// Cancel all watches
-func (values *watches) Cancel() {
-	if values.endpointCancel != nil {
-		values.endpointCancel()
-	}
-	if values.clusterCancel != nil {
-		values.clusterCancel()
-	}
-	if values.routeCancel != nil {
-		values.routeCancel()
-	}
-	if values.listenerCancel != nil {
-		values.listenerCancel()
-	}
-	if values.secretCancel != nil {
-		values.secretCancel()
-	}
-	if values.runtimeCancel != nil {
-		values.runtimeCancel()
-	}
-	for _, cancel := range values.cancellations {
-		if cancel != nil {
-			cancel()
-		}
-	}
-	for _, terminate := range values.terminations {
-		close(terminate)
-	}
+func NewServerAdvanced(restServer rest.Server, sotwServer sotw.Server) Server {
+	return &server{rest: restServer, sotw: sotwServer}
 }
 
-func createResponse(resp cache.Response, typeURL string) (*discovery.DiscoveryResponse, error) {
-	if resp == nil {
-		return nil, errors.New("missing response")
-	}
-
-	marshalledResponse, err := resp.GetDiscoveryResponse()
-	if err != nil {
-		return nil, err
-	}
-
-	return marshalledResponse, nil
-}
-
-// process handles a bi-di stream request
-func (s *server) process(stream Stream, reqCh <-chan *discovery.DiscoveryRequest, defaultTypeURL string) error {
```

Comment:
Can you do git mv v2/server.go to v2/sotw/server.go? It's very difficult to see what changed with the code.

Reply:
I can try, but I really doubt it would help. Basically v2/server.go was split into sotw/v2/server.go and rest/v2/server.go without *code* changes, only methods moved from struct in v2/server.go to implementations of corresponding protocol.

Reply:
I think it would help. sotw has over 300 lines of code, while rest is just 50. It'd help resolving merge conflicts (if any happen) and preserve git history.

Reply:
Tried doing that. Realized that it is not about rest/server.go, it's about the stuff that in current PR is left in original server/server.go - CallbackFuncs, envoy/api methods/etc.

Reply:
So we either save history for server/server.go or sotw/server.go

Reply:
https://github.com/envoyproxy/go-control-plane/pull/350 another PR, we should chose which to leave :)

Reply:
@kyessenov ping